lib/semantics/check-omp-structure.cc

```diff
@@ -14,9 +14,82 @@
 
 #include "check-omp-structure.h"
 #include "../parser/parse-tree.h"
+#include <variant>
 
 namespace Fortran::semantics {
 
+bool OmpStructureChecker::HasInvalidCloselyNestedRegion(
+    const parser::CharBlock &source, const OmpDirectiveEnumType &set) {
+  if (ompContext_.size() && set.test(GetOmpContext().currentDirectiveEnum)) {
+    context_.Say(source,
+        "A worksharing region may not be closely nested inside a "
+        "worksharing, explicit task, taskloop, critical, ordered, atomic, or "
+        "master region."_err_en_US);
+    return true;
+  }
+  return false;
+}
+
+void OmpStructureChecker::CheckAllowed(const OmpClauseEnum &type) {
+  if (ompContext_.empty()) {
+    return;
+  }
+  bool f{true};
+  if (!GetOmpContext().allowedClauses.test(type)) {
+    context_.Say(GetOmpContext().currentClauseSource,
+        "'%s' clause is not allowed on the %s directive"_err_en_US,
+        parser::ToUpperCaseLetters(EnumToString(type)),
```

Comment:
The enumerators are all uppercase, right? So you don't need the call to `ToUpperCaseLetters`.